core/src/rpc.rs

```diff
@@ -126,24 +126,19 @@ impl JsonRpcRequestProcessor {
                     .unwrap()
                     .highest_confirmed_slot();
                 debug!("RPC using confirmed slot: {:?}", slot);
-                Ok(r_bank_forks.get(slot).cloned().unwrap())
+                slot
             }
             CommitmentLevel::Max => {
-                let cluster_root = self
+                let slot = self
                     .block_commitment_cache
                     .read()
                     .unwrap()
                     .largest_confirmed_root();
-                debug!("RPC using block: {:?}", cluster_root);
-                r_bank_forks.get(cluster_root).cloned().ok_or_else(|| {
-                    RpcCustomError::NonexistentClusterRoot {
-                        cluster_root,
-                        node_root: r_bank_forks.root(),
-                    }
-                    .into()
-                })
+                debug!("RPC using block: {:?}", slot);
+                slot
             }
-        }
+        };
+        r_bank_forks.get(slot).cloned().unwrap()
```

Comment:
@garious , looks to me like this will panic when a cluster root doesn't exist on a node, instead of returning the rpc error 😬

Reply:
So there was no test for this and my best guess was that the genesis block was the default cluster root. Is it not? Should it be?

Reply:
Technically, slot 0 is the default cluster root when the largest cluster root is unknown, yes. But the genesis Bank isn't retained in BankForks forever. Should it be?

This case also comes up when a node knows what the largest cluster root is, but that slot does not exist on that node. We've seen this come up on restarts (https://github.com/solana-labs/solana/pull/10061) but there is a workaround in place. Right now, I think this case would only happen if a node is way off on its own fork, so arguably something is going wrong with the node or consensus, but I don't think panicking on an rpc request is the right behavior.

Reply:
A panic is only appropriate if it indicates a bug in the code. In this case, I think it does. It sounds like BankForks should always retain a largest cluster root. If there's a problem on restarts, I wonder, what information is missing from the snapshot, such that BankForks isn't restored?